drivers/include/soft_spi.h

```diff
@@ -0,0 +1,286 @@
+
+#include "periph/gpio.h"
+#include "periph/spi.h"
+#include "mutex.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Default SPI device access macro
+ */
+#ifndef SOFT_SPI_DEV
+#define SOFT_SPI_DEV(x)      (x)
+#endif
+
+/**
+ * @brief   Define global value for undefined SPI device
+ */
+#ifndef SOFT_SPI_UNDEF
+#define SOFT_SPI_UNDEF       (UINT_MAX)
+#endif
+
+/**
+ * @brief   Define value for unused CS line
+ */
+#ifndef SOFT_SPI_CS_UNDEF
+#define SOFT_SPI_CS_UNDEF    (GPIO_UNDEF)
+#endif
+
+/**
+ * @brief   Default type for SPI devices
+ */
+typedef unsigned int soft_spi_t;
+
+/**
+ * @brief   Chip select pin type overlaps with gpio_t so it can be casted to
+ *          this
+ */
+#ifndef HAVE_SOFT_SPI_CS_T
+typedef gpio_t soft_spi_cs_t;
+#endif
+
+/**
+ * @brief   Status codes used by the SPI driver interface
+ */
+enum {
+    SOFT_SPI_OK          =  0,   /**< everything went as planned */
+    SOFT_SPI_NODEV       = -1,   /**< invalid SPI bus specified */
+    SOFT_SPI_NOCS        = -2,   /**< invalid chip select line specified */
+    SOFT_SPI_NOMODE      = -3,   /**< selected mode is not supported */
+    SOFT_SPI_NOCLK       = -4    /**< selected clock value is not supported */
+};
+
+/**
+ * @brief   Available SPI modes, defining the configuration of clock polarity
+ *          and clock phase
+ *
+ * RIOT is using the mode numbers as commonly defined by most vendors
+ * (https://en.wikipedia.org/wiki/Serial_Peripheral_Interface_Bus#Mode_numbers):
+ *
+ * - MODE_0: CPOL=0, CPHA=0 - The first data bit is sampled by the receiver on
+ *           the first SCK rising SCK edge (this mode is used most often).
+ * - MODE_1: CPOL=0, CPHA=1 - The first data bit is sampled by the receiver on
+ *           the second rising SCK edge.
+ * - MODE_2: CPOL=1, CPHA=0 - The first data bit is sampled by the receiver on
+ *           the first falling SCK edge.
+ * - MODE_3: CPOL=1, CPHA=1 - The first data bit is sampled by the receiver on
+ *           the second falling SCK edge.
+ */
+#ifndef HAVE_SOFT_SPI_MODE_T
+typedef enum {
+    SOFT_SPI_MODE_0 = 0,         /**< CPOL=0, CPHA=0 */
+    SOFT_SPI_MODE_1,             /**< CPOL=0, CPHA=1 */
+    SOFT_SPI_MODE_2,             /**< CPOL=1, CPHA=0 */
+    SOFT_SPI_MODE_3              /**< CPOL=1, CPHA=1 */
+} soft_spi_mode_t;
+#endif
+
+/**
+ * @brief   Available SPI clock speeds
+ *
+ * The actual speed of the bus can vary to some extend, as the combination of
+ * CPU clock and available prescaler values on certain platforms may not make
+ * the exact values possible.
+ */
+#ifndef HAVE_SOFT_SPI_CLK_T
+typedef enum {
+    SOFT_SPI_CLK_100KHZ = 0,     /**< drive the SPI bus with 100KHz */
+    SOFT_SPI_CLK_400KHZ,         /**< drive the SPI bus with 400KHz */
+    SOFT_SPI_CLK_1MHZ,           /**< drive the SPI bus with 1MHz */
+    SOFT_SPI_CLK_5MHZ,           /**< drive the SPI bus with 5MHz */
+    SOFT_SPI_CLK_10MHZ           /**< drive the SPI bus with 10MHz */
+} soft_spi_clk_t;
```

Comment:
I think we agreed on running as fast as possible, why did you introduce speed types?

Reply:
to have similar signature with the function of hardware spi, the function still have the clock speed parameter.
also if programmer do a search and replace to adapt the hardware spi using code to use soft spi, the don't have to manually change the clock speed value/constants.
speed may also be implemented later in time when needed although right now the parameter is ignored in the transmission